### PR TITLE
Feat: Adding Linear Algebra Dot operation support

### PR DESCRIPTION
**Pull request by SwayamInSync**
This PR contributes as follows:
- Ship the `dot` method within package that supports following operations
	- vector-vector dot product
	- matrix-vector dot product
	- matrix-matrix multiplication
- Optimized Linear Algebra ops supported by the [QBLAS](https://github.com/SwayamInSync/QBLAS) on x86-64 Linux and ARM machine. On windows it fallbacks to naive implementation due to QBLAS incompatibility with MSVC
- Test Suite to validate dot products between inputs

Images below are the performance comparison
 
quadblas-x86-64-96-cores

> Machine: x86-64 Linux with 96 cores

quadblas-ARM-8-core

> Machine: MacOS-Silicon (ARM) with 8 cores

To compile without QBLAS set `DISABLE_QUADBLAS` as `CFLAGS` and `CXXFLAGS`

### CODE REVIEW

**Comment by juntyr** [comment]
Would the new functionality only be accessed through the dot function or is there a way to call that automatically when using numpy dot, matmul, etc?

**Comment by ngoldbaum** [comment]
(unless I'm missing an existing hook - @seberg might know better)

**Comment by SwayamInSync** [comment]
I see, so if anyways we create a slot, the implementation need to be in done for the quaddtype.
Now with @seberg 's comment I remember trying `NPY_DT_PyArray_ArrFuncs_dotfunc` but it also wasn't supported.

The current `dot` implementation (shipped with package) handles the vector-vector, gemv or gemm as per the input dimensions.

It would be good to have a slot parallel to `np.dot` (like it also picks the operation based on input dimensions)

**Comment by SwayamInSync** [comment]
> Ahh forgot to update the general CI, or should we remove the `quaddtype` from there, given that `build_wheels.yml` has those same checks in its CI with more strict and on all platforms

@ngoldbaum thoughts on this?

**Comment by SwayamInSync** [comment]
As per @seberg 's suggestion, this now implements `numpy.matmul` ufunc. 

Some extra stuff includes:
- Refactor of umath
- Added a `release_tracker.md` to track the progress (pinned all the available numpy ufuncs and what we support right now)

@ngoldbaum please take a look and let me know if anything needs a fix

File: quaddtype/release_tracker.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Plan for `numpy-quaddtype` v1.0.0
+
+| ufunc name    | Added |
```
Comment (juntyr) [inline quaddtype/release_tracker.md:3]: Could you add a column for which ufuncs have tests against float and np.float64 that ensure the correct handling of edge cases (NaN, inf, -0.0, points in [0, 2pi], ...)? I'd like to ensure that by publication we can guarantee that all ufuncs comply with the numpy requirements (see e.g. my issue about mod being incorrect right now)

Comment (SwayamInSync) [reply quaddtype/release_tracker.md:3]: I see, sure will add the corresponding column as well

**Comment by ngoldbaum** [comment]
This is a lot of code! I'll try to give this a once-over but I'm probably not going to have the bandwidth to go over this with a fine-toothed comb. @juntyr I'd appreciate it if you could also do a round of code review. Don't be afraid to ask questions if anything is confusing to you.

**Comment by SwayamInSync** [comment]
> This is a lot of code! I'll try to give this a once-over but I'm probably not going to have the bandwidth to go over this with a fine-toothed comb. @juntyr I'd appreciate it if you could also do a round of code review. Don't be afraid to ask questions if anything is confusing to you.

Yeah the refactor of umath made it big (but those codes are already reviewed), here is what's new
- umath/matmul.cpp & .h files
- quadblas_interface.cpp and h files
- tests/test_dot.py
- GitHub CI files

File: quaddtype/numpy_quaddtype/src/umath/matmul.cpp
[the remark this answers is not in the document]
Comment (SwayamInSync) [reply quaddtype/numpy_quaddtype/src/umath/matmul.cpp:1]: Yup

File: quaddtype/numpy_quaddtype/src/umath/matmul.cpp [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+                                npy_intp *NPY_UNUSED(view_offset))
+{
+    QuadPrecDTypeObject *descr_in1 = (QuadPrecDTypeObject *)given_descrs[0];
+    QuadPrecDTypeObject *descr_in2 = (QuadPrecDTypeObject *)given_descrs[1];
+
+    // QBLAS only supports SLEEF backend
+    if (descr_in1->backend != BACKEND_SLEEF || descr_in2->backend != BACKEND_SLEEF) {
+        PyErr_SetString(PyExc_NotImplementedError,
+                        "QBLAS-accelerated matmul only supports SLEEF backend. "
+                        "Other backends are not supported with QBLAS.");
```
[the remark this answers is not in the document]
Comment (SwayamInSync) [reply quaddtype/numpy_quaddtype/src/umath/matmul.cpp:39]: This is QBLAS intrinsics, for now I implemented it using SLEEF based vectorized operations. 
For longdouble I think OpenBLAS and this one I recently found https://github.com/nakatamaho/mplapack are better alternatives.
Although if demanded then we can move towards that direction

Will add the error message to open issue on QBLAS for this rather than here

File: quaddtype/tests/test_dot.py [hunk 3]
```diff
@@ -0,0 +1,796 @@
+import pytest
+import numpy as np
+from numpy_quaddtype import QuadPrecision, QuadPrecDType
+
+
+# ================================================================================
+# UTILITIES
+# ================================================================================
```
[the remark this answers is not in the document]
Comment (SwayamInSync) [reply quaddtype/tests/test_dot.py:8]: got it

**Comment by SwayamInSync** [comment]
Thanks @ngoldbaum will apply all the changes in next commit :) and so sorry this one got hectic

File: quaddtype/tests/test_utils.py
Comment (ngoldbaum) [inline quaddtype/tests/test_utils.py:1]: This should probably be renamed to e.g. `utils.py`, otherwise if you ever put a function in here that has a name beginning with `test_`, pytest will execute it as a test.

**Review by ngoldbaum — APPROVED** [review]
Just one minor comment - go ahead and self-merge!

I didn't test the windows instructions but we can always fix issues later and I think what's in here is a significant improvement over the nothing we had before :)

**Comment by SwayamInSync** [comment]
> I didn't test the windows instructions but we can always fix issues later and I think what's in here is a significant improvement over the nothing we had before :)

I wrote these parallel to windows CI and tested them on windows-11 PC
Remaining all ufuncs (from tracker) are easily doable @juntyr  tackling some, so I'll most probably start the branch for auto documentation and support for high-endianess systems
